Fix stalled events

mattgibson opened the pull request:
Hi Dan.

Here's the re-done master branch version of the patch I sent for 2.2. Not sure if this is the workflow you wan to follow for 2.2/2.3 versions, but let me know if you'd prefere something different. Original (messy rebase) message follows. 

I found a problem with our events queue getting stuck, so I've had a poke about and found that sometimes, there is an error from Turnitin e.g. user has a duplicate email, but it is not recorded in the plagiarism_turnitin_files status field. There was also a problem where the turnitin_send_file() function would return null in that case, which was causing false to get sent back to the event handler, which then stopped processing, on the assumption that Turnitin was down.

This patch fixes a few other things too. Full list:

turnitin_send_file() returns boolean all the time instead of either nothing or the XML (which is never used).
Problems e.g. cannot enrol in class also now return true so that the events clear.
The event handler has updated code so that files that are found to be already sent return true always, which takes the event out of the queue and lets the cron job take over.
There is now an option to send an array of stored files along with the event.
Fixed a problem where whitespace in the list of codes to resent was stopping them from working.
Added some PHPDoc comments.
Status codes are always saved so we can see the code if it gets stuck halfway.
Turnitin session is ended cleanly even if the process gets stuck.
Problem with undefined variable in the sanity checks in turnitin_send_file() is gone.
I'm afraid these aren't very cleanly separated into different commits, but hopefully that doesn't matter.

Matt

Commit:
Improved flexibility for module files

mattgibson committed:
Added checks for files that have already been sent. Without these, the file is not sent, which then returns false and stops the processing of the event queue.

mattgibson committed:
turnitin_send_file returns boolean explicitly

mattgibson committed:
Reordered the sanity checks to avoid undefined variables

mattgibson committed:
Comments

mattgibson committed:
Reorganised file upload function and made it save status codes even on failure as well as have mor detiled debugging messages.

mattgibson committed:
Fixed issue with whitespace around status codes.

danmarsden commented on lib.php:
```diff
@@ -529,6 +533,21 @@ public function event_handler($eventdata) {
                 return false;
             }
 
+			if($files = $eventdata->files) {
```
HI Matt - we're going to deprecate $eventdata->files soon which is why it was removed from the master code - do you have custom code that uses/relies on this?

mattgibson replied:
Yeah, we have, but can re-do it. What's the preferred method to replace it with?

danmarsden commented on lib.php:
```diff
@@ -544,14 +563,15 @@ public function event_handler($eventdata) {
-                        foreach ($files as $file) {
-                            $fileresult = false;
-                            //TODO: need to check if this file has already been sent! - possible that the file was sent before draft submit was set.
+					if ($files = $fs->get_area_files($modulecontext->id, 'mod_'.$modulename, 'submission', $submission->id, "timemodified", false)) {
```
tabs look a little weird here?

danmarsden commented:
looks pretty good to me - I've neglected the Moodle 2 version of this code a bit since the announcement of the new version but it's taking a bit longer than expected for a new plugin to be available.

just a couple of comments there - we plan to remove eventdata->files from the events in core pretty soon - mainly as they aren't a very reliable way of passing the files around - using the hashes is a lot better. Do you have custom plugins that trigger those events but don't pass the hashes?

mattgibson commented:
Ah. Just saw your later comment. I'll rewrite the files stuff to use pathname hashes and then take that bit out again. Don't know how those tabs got in there, but they'll be gone too.

danmarsden commented:
cool - yeah, turned out passing full file objects into events was fragile - esp if the events were handled without including filelib. :-( - thanks!!!

mattgibson committed:
Fixed indentation - convert to spaces

mattgibson committed:
Updated phpDoc comments

mattgibson committed:
Removed old code to deal with attached files

mattgibson committed:
Made sure $result is always set

mattgibson committed:
Moved assignment out of condition to avoid IDE error

mattgibson commented:
OK, I think that fixes the two issues above, plus I added basic auto-generated PHPDoc comments to the functions and fixed a couple of minor IDE complaints as well.

danmarsden commented:
looks good - shame I can't auto merge this using github :-( - will merge and let you know!

danmarsden commented:
fixed a couple of small conflicts and merged - thanks heaps! - the code has been a bit neglected! - nice to have it operating a bit more reliably!
